File: 탑.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>
using namespace std;

vector<int> solution(vector<int> heights) {
	vector<int> answer;
	stack<int> temp;

	for (int i = heights.size() - 1; i > 0; i--) {
		for (int j = i - 1; j >= 0; j--) {
			if (heights[i] < heights[j]) {
				temp.push(j + 1);
				break;
			}
			else if (j == 0) {
				temp.push(0);
				break;
			}
		}
	}
	temp.push(0);
	while (!temp.empty()) {
		answer.push_back(temp.top());
		temp.pop();
	}

	return answer;
}

File: STACK/탑.py
import sys
from collections import deque
sys.stdin = open("input.txt", "r")

def solution():

    n = int(input())
    tops = list(map(int, input().split()))

    stack = deque()
    answer = [0] * n
    for i, top in enumerate(tops):
        while stack:
            if top > stack[-1][1]:
                stack.pop()
            else:
                break
        if not stack:
            answer[i] = '0'
        else:
            answer[i] = str(stack[-1][0])

        stack.append([i+1, top])

    print(' '.join(answer))





def main():

    solution()

if __name__ == "__main__":
    main()